compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Validate.hs

```diff
@@ -0,0 +1,71 @@
+-- Copyright (c) 2020 The DAML Authors. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module DA.Daml.Compiler.Validate (validateDar) where
+
+import Control.Monad (forM_)
+import qualified "zip-archive" Codec.Archive.Zip as ZipArchive
+import qualified DA.Daml.LF.Ast as LF
+import qualified DA.Daml.LF.Proto3.Archive as Archive
+import qualified Data.ByteString.Lazy as BSL
+import qualified Data.NameMap as NM
+
+import DA.Daml.Compiler.ExtractDar (extractDar,ExtractedDar(..))
+import DA.Daml.LF.Ast.Version
+import DA.Daml.LF.Ast.World (initWorld)
+import DA.Pretty (renderPretty)
+import qualified DA.Daml.LF.TypeChecker as TC (checkModule)
+import qualified DA.Daml.LF.TypeChecker.Error as TC
+
+-- | Monad for performing Validation
+newtype Validation a = Validation (Either ValidationError a) deriving (Functor,Applicative,Monad)
+
+data ValidationError
+  = VeArchiveError FilePath Archive.ArchiveError
+  | VeTypeError TC.Error
+
+validationError :: ValidationError -> Validation a
+validationError = Validation . Left
+
+-- | Run a Validation, throwing an error when the validation fails
+runValidation :: Validation a -> a
+runValidation = \case
+  Validation (Left e) -> error (show e)
```

Comment:
Given that you are only calling this in `IO`, this should use `fail` (or maybe even better `errorIO` which has a slightly less stupid error message) and change the return type. That both makes it more explicit that it can fail and it avoids issues due to laziness. e.g. `_ <- return $ error "foobar"` is not going to blow up whereas `_ <- errorIO "foobar"` will definitely blow up.

Reply:
I used `fail`. Where is `errorIO`?

Reply:
`Control.Exception.Extra`, see https://hoogle.haskell.org/?hoogle=errorIO%20is%3Aexact. The reason why `fail` is a bit annoying is that it prefixes the error message with `user error` which always looks a bit awkward.

Reply:
using `errorIO` now